rust/log_viewer/src/main.rs

```diff
@@ -0,0 +1,246 @@
+// consistent brightness.
+//
+// In order to sample distinct points we use quasi-monte color methods, specifically
+// low-discrepancy sequences. These give us consistently spaced out values in any dimensions
+// and the minimum area/volume between anytime sampled points is inversely related to the
+// number of points sampled. So if we only sample a few points, our colors will be very
+// distinct, as we sample more and more points they will get closer and closer togethor but
+// they should do a quite good job of staying as far apart as possible. We're using an additive
+// recurence which does quite a good job for this use case and is very simple.
+fn generate_distinct_colors() -> impl Iterator<Item = Color> {
+    //let params = JoeKuoD6::minimal();
+    //let seq = Sobol::<f32>::new(2, &params);
+    AdditiveRec {
+        x: 0.0f32,
+        y: 0.0f32,
+        z: 0.0f32,
+    }
+    .map(|point| {
+        // Vary luminosity from 60 to 90. I just played around with it
+        // and found this to be a good range
+        let l = point.0 * 30.0f32 + 60.0f32;
+        //let l = 75.0f32;
+        // The "chroma" of a CIELAB color is sort of like the magnitude
+        // of the A and B components. As chroma gets high the colors start
+        // to look a bit samey. More formally the perceptual rate of change
+        // is low. So instead of using the full -128 to 127 range we use
+        // a -100 to 100 range which I found to work a bit better in practice.
+        let a = 255.0f32 * point.1 - 128.0f32;
+        let b = 255.0f32 * point.2 - 128.0f32;
+        // Add a new color that should be highly distinct from
+        // all the other colors and still have decent brightness
+        let cielab = Lab::new(l, a, b);
+
+        // Going from CIELAB to rgb is complicated but luckily pallate does
+        // that for us.
+        let rgb: Rgb<Srgb, f32> = Rgb::from_color_unclamped(cielab);
+        let rgb: Rgb<Srgb, u8> = rgb.into_linear().into_encoding();
+
+        // Lastly since we're outputting to a terminal we can do a simple conversion
+        // to crossterm's color type
+        Color::Rgb {
+            r: rgb.red,
+            g: rgb.green,
+            b: rgb.blue,
+        }
+    })
+}
+
+fn neg_one() -> i32 {
+    -1
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Event {
+    time: Option<NaiveDateTime>,
+    #[serde(
+        default = "neg_one",
+        deserialize_with = "deserialize_number_from_string"
+    )]
+    pid: i32,
+    level: Option<String>,
+    message: Option<String>,
+    #[serde(flatten)]
+    extra: HashMap<String, Value>,
+}
+
+fn get_log_level_color(level: &Option<String>) -> Color {
+    let Some(level) = level
+    else {
+      return Color::White;
+    };
+
+    match level {
+        x if x == "info" => Color::Grey,
+        x if x == "warning" => Color::Yellow,
+        x if x == "error" => Color::Red,
+        _ => Color::White,
```

Comment:
This seems like a really weird way to spell this out. does 

```

    match level {
        "info" => Color::Grey,
        "warning" => Color::Yellow,
        "error" => Color::Red,
        _ => Color::White,
```

not work?

Reply:
I tried that first and there was an issue with it. I didn't look into the specifics of how rust does that and instead just did this because I knew it would work 🤷

Reply:
hmmmmmm we should probably look into this or at least have an idea _why_ it doesn't work. Doesn't have to be this PR but I'm not super happy with working around compiler errors without knowing why

Reply:
I just needed to add `level.as_str()` all fixed now